tsdb/wlog: Reuse compression decode buffers in LiveReader to reduce allocations

aimuz opened the pull request:
#### Which issue(s) does the PR fix:

This PR addresses high memory allocations observed during WAL checkpoint reading, partially addressing #13174.

In `tsdb/wlog/watcher.go`, the `readCheckpoint` method iterates over WAL segments and creates a new `LiveReader` for each one. Previously, `LiveReader` initialized a fresh `compression.DecodeBuffer` every time. Consequently, the internal buffer passed to `snappy.Decode` always had zero capacity. 

Because the destination buffer was insufficient, `snappy.Decode` was forced to allocate a new byte slice for every decompressed record. In the context of a loop processing many segments, this resulted in excessive memory allocations and GC pressure.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-notes block below.
Otherwise, please describe what should be mentioned in the CHANGELOG. Use the following prefixes:
[FEATURE] [ENHANCEMENT] [PERF] [BUGFIX] [SECURITY] [CHANGE]
Refer to the existing CHANGELOG for inspiration:  https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md
If you need help formulating your entries, consult the reviewer(s).
-->
```release-notes
NONE
```

aimuz committed:
tsdb/wlog: Reuse compression decode buffers in LiveReader

Currently, the readCheckpoint method in the WAL watcher loops over segments
and creates a NewLiveReader for each one. The LiveReader struct previously
allocated a new compression.DecodeBuffer upon initialization. This caused
high memory allocations (specifically for snappy.Decode) because these
buffers were not being reused across the loop iterations.

This commit introduces a global pool for DecodeBuffers using util/zeropool.
The LiveReader now retrieves a buffer from the pool only when needed in
buildRecord and returns it immediately via defer. This allows the heavy
decompression buffers to be reused across multiple LiveReader instances,
reducing memory pressure during WAL checkpointing.

Updates #13174

Signed-off-by: aimuz <mr.imuz@gmail.com>

aimuz commented:
Other issues related to #13174 are still being resolved.

codesome reviewed (commented):
Can you run some benchmarks to see the memory allocation gains? (if there isn't one, may need to write one)

aimuz committed:
add benchmark

Signed-off-by: aimuz <mr.imuz@gmail.com>

aimuz committed:
license header

Signed-off-by: aimuz <mr.imuz@gmail.com>

GiedriusS commented on tsdb/wlog/live_reader.go:
```diff
@@ -105,6 +105,14 @@ type LiveReader struct {
 
+// Close closes the reader and returns the decode buffer to the pool.
+func (r *LiveReader) Close() {
+	if r.decBuf != nil {
```
When can this be nil?

GiedriusS commented on tsdb/wlog/benchmark_test.go:
```diff
@@ -0,0 +1,72 @@
+		// Simulate the loop in Watcher.readCheckpoint or similar where a new LiveReader is created for the segment.
+		r := NewLiveReader(logger, metrics, bytes.NewReader(segData))
+		for r.Next() {
+			_ = r.Record()
```
Should we check that it returns something?

GiedriusS reviewed (commented):
Why benchmarks aren't showing any diff? 🤔

bboreham commented:
> My current implementation has issues, and I am reworking this problem.

I didn't see any follow-up to say you had finished reworking.

You posted benchmarks which show no improvement.

I'll leave it until there is more reason to spend time reviewing.

aimuz commented:
Yes, there is an issue with the current benchmark. I'll re-upload the benchmark code in a moment.

aimuz commented:
new Benchmark

```
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/tsdb/wlog
cpu: Apple M4
                                            │   old.txt   │              new.txt               │
                                            │   sec/op    │   sec/op     vs base               │
LiveReader_DecodeLargeRecord/readers=1-10     1.339m ± 1%   1.283m ± 0%  -4.21% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=16-10    21.64m ± 0%   20.60m ± 1%  -4.77% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=128-10   174.4m ± 1%   165.8m ± 0%  -4.96% (p=0.000 n=10)
geomean                                       17.16m        16.36m       -4.65%

                                            │   old.txt    │               new.txt               │
                                            │     B/s      │     B/s       vs base               │
LiveReader_DecodeLargeRecord/readers=1-10     5.833Gi ± 1%   6.089Gi ± 0%  +4.40% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=16-10    5.778Gi ± 0%   6.067Gi ± 1%  +5.01% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=128-10   5.734Gi ± 1%   6.033Gi ± 0%  +5.22% (p=0.000 n=10)
geomean                                       5.781Gi        6.063Gi       +4.87%

                                            │    old.txt    │               new.txt                │
                                            │     B/op      │     B/op      vs base                │
LiveReader_DecodeLargeRecord/readers=1-10     309.51Ki ± 0%   53.63Ki ± 1%  -82.67% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=16-10    4952.0Ki ± 0%   858.1Ki ± 1%  -82.67% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=128-10   38.688Mi ± 0%   6.706Mi ± 0%  -82.67% (p=0.000 n=10)
geomean                                        3.838Mi        681.1Ki       -82.67%

                                            │   old.txt   │              new.txt               │
                                            │  allocs/op  │ allocs/op   vs base                │
LiveReader_DecodeLargeRecord/readers=1-10      6.000 ± 0%   3.000 ± 0%  -50.00% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=16-10    101.00 ± 0%   49.00 ± 0%  -51.49% (p=0.000 n=10)
LiveReader_DecodeLargeRecord/readers=128-10    813.0 ± 0%   397.0 ± 0%  -51.17% (p=0.000 n=10)
geomean                                        78.98        38.79       -50.89%
```

aimuz committed:
tsdb/wlog: improve LiveReader decode benchmark

Replace the existing LiveReader segment benchmark with one that more
directly exercises the allocation pattern fixed by decode buffer reuse.

The previous benchmark used small records and a single reader, which made
it difficult to observe the cost of repeatedly allocating decoded output
buffers. The new benchmark writes large, highly-compressible records and
runs the same WAL data through 1, 16, and 128 LiveReader instances.

This keeps the benchmark focused on LiveReader behavior while making the
allocation reduction visible for repeated reader creation.

Also remove the redundant nil check in LiveReader.Close now that the
decode buffer is always returned through PutDecodeBuffer.

Signed-off-by: aimuz <mr.imuz@gmail.com>

aimuz commented:
@GiedriusS @bboreham 

Hi, I've mainly adjusted the benchmark code. It's ready and awaiting review.

GiedriusS commented:
I'm surprised to see increase in B/s. Maybe you tried looking into why that happens?
